packages/api/src/clickhouse/index.ts

```diff
@@ -904,6 +912,588 @@ export const getMetricsChart = async ({
+    throw new Error(
+      'Field is required for all aggregation functions except Count',
+    );
+  }
+
+  const selectField =
+    field != null
+      ? buildSearchColumnName(propertyTypeMappingsModel.get(field), field)
+      : '';
+
+  const hasGroupBy = groupBy != '' && groupBy != null;
+  const isCountFn = aggFn === AggFn.Count;
+  const groupByField =
+    hasGroupBy &&
+    buildSearchColumnName(propertyTypeMappingsModel.get(groupBy), groupBy);
+
+  const serializer = new SQLSerializer(propertyTypeMappingsModel);
+
+  const label = SqlString.escape(`${aggFn}(${field})`);
+
+  const selectClause = [
+    isCountFn
+      ? 'toFloat64(count()) as data'
+      : aggFn === AggFn.Sum
+      ? `sum(${selectField}) as data`
+      : aggFn === AggFn.Avg
+      ? `avg(${selectField}) as data`
+      : aggFn === AggFn.Max
+      ? `max(${selectField}) as data`
+      : aggFn === AggFn.Min
+      ? `min(${selectField}) as data`
+      : aggFn === AggFn.CountDistinct
+      ? `count(distinct ${selectField}) as data`
+      : `quantile(${
+          aggFn === AggFn.P50
+            ? '0.5'
+            : aggFn === AggFn.P90
+            ? '0.90'
+            : aggFn === AggFn.P95
+            ? '0.95'
+            : '0.99'
+        })(${selectField}) as data`,
+    granularity != null
+      ? `toUnixTimestamp(toStartOfInterval(timestamp, INTERVAL ${granularity})) as ts_bucket`
+      : "'0' as ts_bucket",
+    groupByField ? `${groupByField} as group` : `'' as group`, // FIXME: should we fallback to use aggFn as group
+    `${label} as label`,
+  ].join(',');
+
+  const groupByClause = `ts_bucket ${groupByField ? `, ${groupByField}` : ''}`;
+
+  const query = SqlString.format(
+    `
+      SELECT ?
+      FROM ??
+      WHERE ? AND (?) ? ?
+      GROUP BY ?
+      ORDER BY ts_bucket ASC
+      ${
+        granularity != null
+          ? `WITH FILL
+        FROM toUnixTimestamp(toStartOfInterval(toDateTime(?), INTERVAL ?))
+        TO toUnixTimestamp(toStartOfInterval(toDateTime(?), INTERVAL ?))
+        STEP ?`
+          : ''
+      }${
+      sortOrder === 'asc' || sortOrder === 'desc' ? `, data ${sortOrder}` : ''
+    }
+    `,
+    [
+      SqlString.raw(selectClause),
+      tableName,
+      buildTeamLogStreamWhereCondition(tableVersion, teamId),
+      SqlString.raw(whereClause),
+      SqlString.raw(
+        !isCountFn && field != null
+          ? ` AND (${await serializer.isNotNull(field, false)})`
+          : '',
+      ),
+      SqlString.raw(
+        hasGroupBy
+          ? ` AND (${await serializer.isNotNull(groupBy, false)})`
+          : '',
+      ),
+      SqlString.raw(groupByClause),
+      ...(granularity != null
+        ? [
+            startTime / 1000,
+            granularity,
+            endTime / 1000,
+            granularity,
+            ms(granularity) / 1000,
+          ]
+        : []),
+    ],
+  );
+
+  return {
+    query,
+    hasGroupBy,
+  };
+};
+
+export const queryMultiSeriesChart = async ({
```

Comment:
I can't really read this one and it hurts my brain. I think a simple test helps (at least I can see what the output query looks like). What I can tell so far is it concats a bunch of CTE and merge them afterward

Reply:
Which lines were too dense to read? If the code can't be read - best that I rewrite the code to make it readable. Fwiw this function literally does just concats a bunch of CTEs and then merges them.

If you want to see the query output - I heard there's a cool tool to observe your systems ;)